File: Trumpet.Backend/Services/Recommendations/RecommendationService.cs
using Microsoft.EntityFrameworkCore;
using Trumpet.Backend.Data;
using Trumpet.Backend.Models;

namespace Trumpet.Backend.Services.Recommendations;

public class RecommendationService : IRecommendationService
{
    private readonly MusicContext _context;

    public RecommendationService(MusicContext context)
    {
        _context = context;
    }

    public async Task<List<Item>> GetRecommendationsAsync(string itemId, int maxResults = 5)
    {
        // 1. Get the tags of the source item
        var sourceTags = await _context.MetadataValues
            .Where(m => m.ItemId == itemId && m.Field == "trumpet.tag")
            .Select(m => m.Value)
            .ToListAsync();

        if (!sourceTags.Any())
        {
            return new List<Item>();
        }

        // 2. Find other items that have these tags
        // We want to count how many matching tags each item has
        var recommendations = await _context.MetadataValues
            // Filter only tag fields
            .Where(m => m.Field == "trumpet.tag" && m.ItemId != itemId && sourceTags.Contains(m.Value))
            // Group by ItemId
            .GroupBy(m => m.ItemId)
            // Select ItemId and Count
            .Select(g => new { ItemId = g.Key, MatchCount = g.Count() })
            // Order by matches descending
            .OrderByDescending(x => x.MatchCount)
            .Take(maxResults)
            .ToListAsync();

        var recommendedItemIds = recommendations.Select(r => r.ItemId).ToList();

        if (!recommendedItemIds.Any())
        {
            return new List<Item>();
        }

        // 3. Fetch the actual Items
        var items = await _context.Items
            .Include(i => i.Metadata)
            .Include(i => i.Collection) // Include collection just in case
            .Where(i => recommendedItemIds.Contains(i.Id))
            .ToListAsync();
            
        // Re-sort in memory because SQL IN clause doesn't guarantee order
        return items
            .OrderByDescending(i => recommendations.First(r => r.ItemId == i.Id).MatchCount)
            .ToList();
    }
}

File: TrumpetPython/services/recommendation_service.py
from typing import List, Optional
from sqlmodel import Session, select, func, and_
from sqlalchemy.orm import selectinload
from models.item import Item
from models.metadata_value import MetadataValue

class RecommendationService:
    def __init__(self, session: Session):
        self.session = session

    async def get_recommendations(self, item_id: str, max_results: int = 5) -> List[Item]:
        # 1. Get the tags of the source item
        source_tags_statement = select(MetadataValue.value).where(
            and_(MetadataValue.item_id == item_id, MetadataValue.field == "trumpet.tag")
        )
        source_tags = self.session.exec(source_tags_statement).all()
        
        if not source_tags:
            return []

        # 2. Find other items that have these tags and count matches
        recommendations_statement = (
            select(MetadataValue.item_id, func.count(MetadataValue.id).label("match_count"))
            .where(
                and_(
                    MetadataValue.field == "trumpet.tag",
                    MetadataValue.item_id != item_id,
                    MetadataValue.value.in_(source_tags)
                )
            )
            .group_by(MetadataValue.item_id)
            .order_by(func.count(MetadataValue.id).desc())
            .limit(max_results)
        )
        
        recommendations = self.session.exec(recommendations_statement).all()
        
        if not recommendations:
            return []

        recommended_item_ids = [r[0] for r in recommendations]
        match_counts = {r[0]: r[1] for r in recommendations}

        # 3. Fetch the actual Items
        items_statement = select(Item).where(
            Item.id.in_(recommended_item_ids)
        ).options(
            selectinload(Item.metadata),
            selectinload(Item.collection)
        )
        
        items = self.session.exec(items_statement).all()
        
        # Sort items based on match counts (since SQL IN doesn't preserve order)
        sorted_items = sorted(items, key=lambda i: match_counts.get(i.id, 0), reverse=True)
        
        return list(sorted_items)
